README.md
![header](./header.png)
# CircleMenu
[![Twitter](https://img.shields.io/badge/Twitter-@Ramotion-blue.svg?style=flat)](http://twitter.com/Ramotion)
[![CocoaPods](https://img.shields.io/cocoapods/p/CircleMenu.svg)](https://cocoapods.org/pods/CircleMenu)
[![CocoaPods](https://img.shields.io/cocoapods/v/CircleMenu.svg)](http://cocoapods.org/pods/CircleMenu)
[![codebeat badge](https://codebeat.co/badges/6f67da5d-c416-4bac-9fb7-c2dc938feedc)](https://codebeat.co/projects/github-com-ramotion-circle-menu)
[![Travis](https://img.shields.io/travis/Ramotion/circle-menu.svg)](https://travis-ci.org/Ramotion/circle-menu)

[shot on dribbble:](https://dribbble.com/shots/2534780-Circle-Menu-Swift-Open-Source)
![preview](./preview.gif)


The [iPhone mockup](https://store.ramotion.com/product/iphone-6-mockups?utm_source=gthb&utm_medium=special&utm_campaign=circle-menu) available [here](https://store.ramotion.com/product/iphone-6-mockups?utm_source=gthb&utm_medium=special&utm_campaign=circle-menu).

## Requirements

- iOS 8.0+
- Xcode 7.2

## Installation

Just add CircleMenuLib folder to your project.

or use [CocoaPods](https://cocoapods.org) with Podfile:

```ruby
pod 'CircleMenu', '~> 1.0.0'
```

## Usage

##### with storyboard

1) Create a new UIButton inheriting from `CircleMenu`

2) Add images for Normal and Selected state

3) Use delegate method to configure buttons

```swift
func circleMenu(circleMenu: CircleMenu, willDisplay button: CircleMenuButton, atIndex: Int)
```

4) Use properties to confiure CircleMenu

```swift
@IBInspectable var buttonsCount: Int = 3
@IBInspectable var duration: Double = 2 // circle animation duration
@IBInspectable var distance: Float = 100 // distance between center button and buttons
```

##### programmatically

```swift
let button = CircleMenu(
  frame: CGRect(x: 200, y: 200, width: 50, height: 50),
  normalIcon:"icon_menu",
  selectedIcon:"icon_close",
  buttonsCount: 4,
  duration: 4,
  distance: 120)
button.delegate = self
button.layer.cornerRadius = button.frame.size.width / 2.0
view.addSubview(button)
```

##### delegate methods

```swift
// configure buttons
optional func circleMenu(circleMenu: CircleMenu, willDisplay button: CircleMenuButton, atIndex: Int)

// call before animation
optional func circleMenu(circleMenu: CircleMenu, buttonWillSelected button: CircleMenuButton, atIndex: Int)

// call after animation
optional func circleMenu(circleMenu: CircleMenu, buttonDidSelected button: CircleMenuButton, atIndex: Int)
```

## Licence

Circle menu is released under the MIT license.
See [LICENSE](./LICENSE) for details.

## About
The project maintained by [app development agency](https://ramotion.com?utm_source=gthb&utm_medium=special&utm_campaign=circle-menu) [Ramotion Inc.](https://ramotion.com?utm_source=gthb&utm_medium=special&utm_campaign=circle-menu)
See our other [open-source projects](https://github.com/ramotion) or [hire](https://ramotion.com?utm_source=gthb&utm_medium=special&utm_campaign=circle-menu) us to design, develop, and grow your product.

[![Twitter URL](https://img.shields.io/twitter/url/http/shields.io.svg?style=social)](https://twitter.com/intent/tweet?text=https://github.com/ramotion/circle-menu)
[![Twitter Follow](https://img.shields.io/twitter/follow/ramotion.svg?style=social)](https://twitter.com/ramotion)


CHANGELOG.md
# Change Log

## [1.0.0](https://github.com/Ramotion/circle-menu/tree/1.0.0) (2016-03-24)
[Full Changelog](https://github.com/Ramotion/circle-menu/compare/0.0.7...1.0.0)

**Closed issues:**

- Add tap gesture recognizer on view to close the menu [\#9](https://github.com/Ramotion/circle-menu/issues/9)

**Merged pull requests:**

- fixed swift 2.2 deprecated warnings in circlemenu.swift [\#10](https://github.com/Ramotion/circle-menu/pull/10) ([sairamkotha](https://github.com/sairamkotha))

## [0.0.7](https://github.com/Ramotion/circle-menu/tree/0.0.7) (2016-03-16)
[Full Changelog](https://github.com/Ramotion/circle-menu/compare/0.0.6...0.0.7)

## [0.0.6](https://github.com/Ramotion/circle-menu/tree/0.0.6) (2016-03-14)
[Full Changelog](https://github.com/Ramotion/circle-menu/compare/0.0.5...0.0.6)

**Closed issues:**

- Problem! [\#7](https://github.com/Ramotion/circle-menu/issues/7)
- Menu icon position [\#6](https://github.com/Ramotion/circle-menu/issues/6)

**Merged pull requests:**

- Fix crash in example [\#3](https://github.com/Ramotion/circle-menu/pull/3) ([skywinder](https://github.com/skywinder))
- Cleaning and docking [\#2](https://github.com/Ramotion/circle-menu/pull/2) ([Skogetroll](https://github.com/Skogetroll))

## [0.0.5](https://github.com/Ramotion/circle-menu/tree/0.0.5) (2016-02-10)
[Full Changelog](https://github.com/Ramotion/circle-menu/compare/0.0.4...0.0.5)

## [0.0.4](https://github.com/Ramotion/circle-menu/tree/0.0.4) (2016-02-09)
[Full Changelog](https://github.com/Ramotion/circle-menu/compare/0.0.3...0.0.4)

## [0.0.3](https://github.com/Ramotion/circle-menu/tree/0.0.3) (2016-02-04)
[Full Changelog](https://github.com/Ramotion/circle-menu/compare/0.0.2...0.0.3)

## [0.0.2](https://github.com/Ramotion/circle-menu/tree/0.0.2) (2016-02-03)
[Full Changelog](https://github.com/Ramotion/circle-menu/compare/0.0.1...0.0.2)

## [0.0.1](https://github.com/Ramotion/circle-menu/tree/0.0.1) (2016-02-03)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

CircleMenu/AppDelegate.swift
//
//  AppDelegate.swift
//  CircleMenu
//
//  Created by Alex K. on 27/01/16.
//  Copyright © 2016 Alex K. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {
    
    var window: UIWindow?
    
    
    func application(
        application: UIApplication,
        didFinishLaunchingWithOptions launchOptions: [NSObject: AnyObject]?) -> Bool {
            return true
    }
    
    func applicationWillResignActive(application: UIApplication) {
        
    }
    
    func applicationDidEnterBackground(application: UIApplication) {
        
    }
    
    func applicationWillEnterForeground(application: UIApplication) {
        
    }
    
    func applicationDidBecomeActive(application: UIApplication) {
        
    }
    
    func applicationWillTerminate(application: UIApplication) {
        
    }
    
    
}


CircleMenuLib/CircleMenu.swift
//
//  CircleMenu.swift
//  ButtonTest
//
// Copyright (c) 18/01/16. Ramotion Inc. (http://ramotion.com)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

import UIKit

// MARK: helpers

@warn_unused_result
public func Init<Type>(value: Type, @noescape block: (object: Type) -> Void) -> Type {
  block(object: value)
  return value
}

// MARK: Protocol

@objc public protocol CircleMenuDelegate {
  
  // don't change button.tag
  optional func circleMenu(circleMenu: CircleMenu, willDisplay button: CircleMenuButton, atIndex: Int)
  
  // call before animation
  optional func circleMenu(circleMenu: CircleMenu, buttonWillSelected button: CircleMenuButton, atIndex: Int)
  
  // call after animation
  optional func circleMenu(circleMenu: CircleMenu, buttonDidSelected button: CircleMenuButton, atIndex: Int)
}

// MARK: CircleMenu
public class CircleMenu: UIButton {
  
  // MARK: properties
  
  @IBInspectable public var buttonsCount: Int = 3
  @IBInspectable public var duration: Double  = 2 // circle animation duration
  @IBInspectable public var distance: Float   = 100 // distance between center button and buttons
  @IBInspectable public var showDelay: Double = 0 // delay between show buttons
  
  
  @IBOutlet weak public var delegate: AnyObject? //CircleMenuDelegate?
  
  var buttons: [CircleMenuButton]?
  
  private var customNormalIconView: UIImageView!
  private var customSelectedIconView: UIImageView!
  
  // MARK: life cycle
  public init(frame: CGRect, normalIcon: String?, selectedIcon: String?, buttonsCount: Int = 3, duration: Double = 2,
    distance: Float = 100) {
      super.init(frame: frame)
      
      if let icon = normalIcon {
        setImage(UIImage(named: icon), forState: .Normal)
      }
      
      if let icon = selectedIcon {
        setImage(UIImage(named: icon), forState: .Selected)
      }
      
      self.buttonsCount = buttonsCount
      self.duration     = duration
      self.distance     = distance
      
      commonInit()
  }
  
  required public init?(coder aDecoder: NSCoder) {
    super.init(coder: aDecoder)
    
    commonInit()
  }
  
  private func commonInit() {
    addActions()
    
    customNormalIconView = addCustomImageView(state: .Normal)
    
    customSelectedIconView = addCustomImageView(state: .Selected)
    if customSelectedIconView != nil {
      customSelectedIconView.alpha = 0
    }
    setImage(UIImage(), forState: .Normal)
    setImage(UIImage(), forState: .Selected)
  }
  
  // MARK: public
  
  public func hideButtons(duration: Double, hideDelay: Double = 0) {
    if buttons == nil {
      return
    }
    
    buttonsAnimationIsShow(isShow: false, duration: duration, hideDelay: hideDelay)

    tapBounceAnimation()
    tapRotatedAnimation(0.3, isSelected: false)
  }

  
  // MARK: create
  
  private func createButtons() -> [CircleMenuButton] {
    var buttons = [CircleMenuButton]()
    
    let step: Float = 360.0 / Float(self.buttonsCount)
    for index in 0..<self.buttonsCount {
      
      let angle: Float = Float(index) * step
      let distance = Float(self.bounds.size.height/2.0)
      let button = Init(CircleMenuButton(size: self.bounds.size, circleMenu: self, distance:distance, angle: angle)) {
          $0.tag = index
          $0.addTarget(self, action: #selector(CircleMenu.buttonHandler(_:)), forControlEvents: UIControlEvents.TouchUpInside)
          $0.alpha = 0
      }
      buttons.append(button)
    }
    return buttons
  }
  
  private func addCustomImageView(state state: UIControlState) -> UIImageView? {
    guard let image = imageForState(state) else {
      return nil
    }
    
    let iconView = Init(UIImageView(image: image)) {
      $0.translatesAutoresizingMaskIntoConstraints = false
      $0.contentMode                               = .Center
      $0.userInteractionEnabled                    = false
    }
    addSubview(iconView)
    
    // added constraints
    iconView.addConstraint(NSLayoutConstraint(item: iconView, attribute: .Height, relatedBy: .Equal, toItem: nil,
      attribute: .Height, multiplier: 1, constant: bounds.size.height))
    
    iconView.addConstraint(NSLayoutConstraint(item: iconView, attribute: .Width, relatedBy: .Equal, toItem: nil,
      attribute: .Width, multiplier: 1, constant: bounds.size.width))
    
    addConstraint(NSLayoutConstraint(item: self, attribute: .CenterX, relatedBy: .Equal, toItem: iconView,
      attribute: .CenterX, multiplier: 1, constant:0))
    
    addConstraint(NSLayoutConstraint(item: self, attribute: .CenterY, relatedBy: .Equal, toItem: iconView,
      attribute: .CenterY, multiplier: 1, constant:0))
    
    return iconView
  }
  
  // MARK: configure
  
  private func addActions() {
    self.addTarget(self, action: #selector(CircleMenu.onTap), forControlEvents: UIControlEvents.TouchUpInside)
  }
  
  // MARK: helpers
  
  public func buttonsIsShown() -> Bool {
    guard let buttons = self.buttons else {
      return false
    }
    
    for button in buttons {
      if button.alpha == 0 {
        return false
      }
    }
    return true
  }
  
  // MARK: actions
  
  func onTap() {
    if buttonsIsShown() == false {
      buttons = createButtons()
    }
    let isShow = !buttonsIsShown()
    let duration  = isShow ? 0.5 : 0.2
    buttonsAnimationIsShow(isShow: isShow, duration: duration)
    
    tapBounceAnimation()
    tapRotatedAnimation(0.3, isSelected: isShow)
  }
  
  func buttonHandler(sender: CircleMenuButton) {
    delegate?.circleMenu?(self, buttonWillSelected: sender, atIndex: sender.tag)
    
    let circle = CircleMenuLoader(radius: CGFloat(distance), strokeWidth: bounds.size.height, circleMenu: self,
      color: sender.backgroundColor)
    
    if let container = sender.container { // rotation animation
      sender.rotationLayerAnimation(container.angleZ + 360, duration: duration)
      container.superview?.bringSubviewToFront(container)
    }
    
    if let aButtons = buttons {
      circle.fillAnimation(duration, startAngle: -90 + Float(360 / aButtons.count) * Float(sender.tag))
      circle.hideAnimation(0.5, delay: duration)
      
      hideCenterButton(duration: 0.3)
      
      buttonsAnimationIsShow(isShow: false, duration: 0, hideDelay: duration)
      showCenterButton(duration: 0.525, delay: duration)
      
      if customNormalIconView != nil && customSelectedIconView != nil {
        let dispatchTime: dispatch_time_t = dispatch_time(
          DISPATCH_TIME_NOW,
          Int64(duration * Double(NSEC_PER_SEC)))
        
        dispatch_after(dispatchTime, dispatch_get_main_queue(), {
          self.delegate?.circleMenu?(self, buttonDidSelected: sender, atIndex: sender.tag)
        })
      }
    }
  }
  
  // MARK: animations
  
  private func buttonsAnimationIsShow(isShow isShow: Bool, duration: Double, hideDelay: Double = 0) {
    guard let buttons = self.buttons else {
      return
    }
    
    let step: Float = 360.0 / Float(self.buttonsCount)
    for index in 0..<self.buttonsCount {
      let button       = buttons[index]
      let angle: Float = Float(index) * step
      if isShow == true {
        delegate?.circleMenu?(self, willDisplay: button, atIndex: index)
        
        button.rotatedZ(angle: angle, animated: false, delay: Double(index) * showDelay)
        button.showAnimation(distance: distance, duration: duration, delay: Double(index) * showDelay)
      } else {
        button.hideAnimation(
          distance: Float(self.bounds.size.height / 2.0),
          duration: duration, delay: hideDelay)
      }
    }
    if isShow == false { // hide buttons and remove
      self.buttons = nil
    }
  }
  
  private func tapBounceAnimation() {
    self.transform = CGAffineTransformMakeScale(0.9, 0.9)
    UIView.animateWithDuration(0.5, delay: 0, usingSpringWithDamping: 0.3, initialSpringVelocity: 5,
      options: UIViewAnimationOptions.CurveLinear,
      animations: { () -> Void in
        self.transform = CGAffineTransformMakeScale(1, 1)
      },
      completion: nil)
  }
  
  private func tapRotatedAnimation(duration: Float, isSelected: Bool) {
    
    let addAnimations: (view: UIImageView, isShow: Bool) -> () = { (view, isShow) in
      var toAngle: Float   = 180.0
      var fromAngle: Float = 0
      var fromScale        = 1.0
      var toScale          = 0.2
      var fromOpacity      = 1
      var toOpacity        = 0
      if isShow == true {
        toAngle     = 0
        fromAngle   = -180
        fromScale   = 0.2
        toScale     = 1.0
        fromOpacity = 0
        toOpacity   = 1
      }
      
      let rotation = Init(CABasicAnimation(keyPath: "transform.rotation")) {
        $0.duration       = NSTimeInterval(duration)
        $0.toValue        = (toAngle.degrees)
        $0.fromValue      = (fromAngle.degrees)
        $0.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
      }
      let fade = Init(CABasicAnimation(keyPath: "opacity")) {
        $0.duration            = NSTimeInterval(duration)
        $0.fromValue           = fromOpacity
        $0.toValue             = toOpacity
        $0.timingFunction      = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
        $0.fillMode            = kCAFillModeForwards
        $0.removedOnCompletion = false
      }
      let scale = Init(CABasicAnimation(keyPath: "transform.scale")) {
        $0.duration       = NSTimeInterval(duration)
        $0.toValue        = toScale
        $0.fromValue      = fromScale
        $0.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
      }
      
      view.layer.addAnimation(rotation, forKey: nil)
      view.layer.addAnimation(fade, forKey: nil)
      view.layer.addAnimation(scale, forKey: nil)
    }
    
    if customNormalIconView != nil && customSelectedIconView != nil {
      addAnimations(view: customNormalIconView, isShow: !isSelected)
      addAnimations(view: customSelectedIconView, isShow: isSelected)
    }
    selected = isSelected
    self.alpha = isSelected ? 0.3 : 1
  }
  
  private func hideCenterButton(duration duration: Double, delay: Double = 0) {
    UIView.animateWithDuration( NSTimeInterval(duration), delay: NSTimeInterval(delay),
      options: UIViewAnimationOptions.CurveEaseOut,
      animations: { () -> Void in
        self.transform = CGAffineTransformMakeScale(0.001, 0.001)
      }, completion: nil)
  }
  
  private func showCenterButton(duration duration: Float, delay: Double) {
    UIView.animateWithDuration( NSTimeInterval(duration), delay: NSTimeInterval(delay), usingSpringWithDamping: 0.78,
      initialSpringVelocity: 0, options: UIViewAnimationOptions.CurveLinear,
      animations: { () -> Void in
        self.transform = CGAffineTransformMakeScale(1, 1)
        self.alpha     = 1
      },
      completion: nil)
    
    let rotation = Init(CASpringAnimation(keyPath: "transform.rotation")) {
      $0.duration        = NSTimeInterval(1.5)
      $0.toValue         = (0)
      $0.fromValue       = (Float(-180).degrees)
      $0.damping         = 10
            $0.initialVelocity = 0
      $0.beginTime       = CACurrentMediaTime() + delay
    }
    let fade = Init(CABasicAnimation(keyPath: "opacity")) {
      $0.duration            = NSTimeInterval(0.01)
      $0.toValue             = 0
      $0.timingFunction      = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
      $0.fillMode            = kCAFillModeForwards
      $0.removedOnCompletion = false
      $0.beginTime           = CACurrentMediaTime() + delay
    }
    let show = Init(CABasicAnimation(keyPath: "opacity")) {
      $0.duration            = NSTimeInterval(duration)
      $0.toValue             = 1
      $0.timingFunction      = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
      $0.fillMode            = kCAFillModeForwards
      $0.removedOnCompletion = false
      $0.beginTime           = CACurrentMediaTime() + delay
    }
    
    if customNormalIconView != nil {
      customNormalIconView.layer.addAnimation(rotation, forKey: nil)
      customNormalIconView.layer.addAnimation(show, forKey: nil)
    }
    
    if customSelectedIconView != nil {
      customSelectedIconView.layer.addAnimation(fade, forKey: nil)
    }
  }
}

// MARK: extension

extension Float {
  var radians: Float {
    return self * (Float(180) / Float(M_PI))
  }
  
  var degrees: Float {
    return self  * Float(M_PI) / 180.0
  }
}

extension UIView {
  
  var angleZ: Float {
    let radians: Float = atan2(Float(self.transform.b), Float(self.transform.a))
    return radians.radians
  }
}


CircleMenuTests/CircleMenuTests.swift
//
//  CircleMenuTests.swift
//  CircleMenuTests
//
//  Created by Alex K. on 27/01/16.
//  Copyright © 2016 Alex K. All rights reserved.
//

import XCTest

@testable import CircleMenu

class CircleMenuTests: XCTestCase {
    
    let buttonsCount = 4
    
    let circleMenu = CircleMenu(
        frame: CGRect(x: 200, y: 200, width: 50, height: 50),
        normalIcon:"icon_menu",
        selectedIcon:"icon_close",
        buttonsCount: 4,
        duration: 4,
        distance: 120)
    
    let view = UIView()
    
    
    override func setUp() {
        super.setUp()
        circleMenu.buttonsCount = buttonsCount
        view.addSubview(circleMenu)
    }
    
    override func tearDown() {
        super.tearDown()
        
    }
    
    func testCircleMenuShowButtons() {
        // given
        XCTAssertNil(circleMenu.buttons, "button doesn't create")
        
        // when button tap
        circleMenu.onTap()
        
        // then
        XCTAssertEqual(circleMenu.buttons?.count, buttonsCount, "button is created")
    }
    
    
    func testCircleMenuHideButtons() {
        // given
        circleMenu.onTap()
        
        // when
        circleMenu.onTap()
        
        // then
        XCTAssertNil(circleMenu.buttons, "button is removed")
    }
    
    
    func testCircleMenuHideButtonsAfterAnimation() {
        // given
        circleMenu.onTap()
        
        // when
        circleMenu.buttonHandler((circleMenu.buttons?.first)!)
        
        // then
        XCTAssertNil(circleMenu.buttons, "button is removed")
    }
}


CircleMenuLib/CircleMenuButton/CircleMenuButton.swift
//
//  CircleMenuButton.swift
//
// Copyright (c) 18/01/16. Ramotion Inc. (http://ramotion.com)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

import UIKit

public class CircleMenuButton: UIButton {
    
    // MARK: properties
    
    public weak var container: UIView?
    
    // MARK: life cycle
    
    init(size: CGSize, circleMenu: CircleMenu, distance: Float, angle: Float = 0) {
        super.init(frame: CGRect(origin: CGPoint(x: 0, y: 0), size: size))
        
        self.backgroundColor = UIColor(colorLiteralRed: 0.79, green: 0.24, blue: 0.27, alpha: 1)
        self.layer.cornerRadius = size.height / 2.0
        
        let aContainer = createContainer(CGSize(width: size.width, height:CGFloat(distance)), circleMenu: circleMenu)
        
        // hack view for rotate
        let view = UIView(frame: CGRect(x: 0, y: 0, width: self.bounds.width, height: self.bounds.height))
        view.backgroundColor = UIColor.clearColor()
        view.addSubview(self)
        //...
        
        aContainer.addSubview(view)
        container = aContainer
        
        view.layer.transform = CATransform3DMakeRotation(-CGFloat(angle.degrees), 0, 0, 1)
        
        self.rotatedZ(angle: angle, animated: false)
    }
    
    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    // MARK: configure
    
    private func createContainer(size: CGSize, circleMenu: CircleMenu) -> UIView {
        
        guard let circleMenuSuperView = circleMenu.superview else {
            fatalError("wront circle menu")
        }
        
        let container = Init(UIView(frame: CGRect(origin: CGPoint(x: 0, y: 0), size: size))) {
            $0.backgroundColor                           = UIColor.clearColor()
            $0.translatesAutoresizingMaskIntoConstraints = false
            $0.layer.anchorPoint                         = CGPoint(x: 0.5, y: 1)
        }
        circleMenuSuperView.insertSubview(container, belowSubview: circleMenu)
        
        // added constraints
        let height = NSLayoutConstraint(item: container,
            attribute: .Height,
            relatedBy: .Equal,
            toItem: nil,
            attribute: .Height,
            multiplier: 1,
            constant: size.height)
        height.identifier = "height"
        container.addConstraint(height)
        
        container.addConstraint(NSLayoutConstraint(item: container,
            attribute: .Width,
            relatedBy: .Equal,
            toItem: nil,
            attribute: .Width,
            multiplier: 1,
            constant: size.width))
        
        circleMenuSuperView.addConstraint(NSLayoutConstraint(item: circleMenu,
            attribute: .CenterX,
            relatedBy: .Equal,
            toItem: container,
            attribute: .CenterX,
            multiplier: 1,
            constant:0))
        
        circleMenuSuperView.addConstraint(NSLayoutConstraint(item: circleMenu,
            attribute: .CenterY,
            relatedBy: .Equal,
            toItem: container,
            attribute: .CenterY,
            multiplier: 1,
            constant:0))
        
        return container
    }
    
    // MARK: public
    
    public func rotatedZ(angle angle: Float, animated: Bool, duration: Double = 0, delay: Double = 0) {
        guard let container = self.container else {
            fatalError("contaner don't create")
        }
        
        let rotateTransform = CATransform3DMakeRotation(CGFloat(angle.degrees), 0, 0, 1)
        if animated {
            UIView.animateWithDuration(
                duration,
                delay: delay,
                options: UIViewAnimationOptions.CurveEaseInOut,
                animations: { () -> Void in
                    container.layer.transform = rotateTransform
                },
                completion: nil)
        } else {
            container.layer.transform = rotateTransform
        }
    }
}

// MARK: Animations

extension CircleMenuButton {
    
    public func showAnimation(distance distance: Float, duration: Double, delay: Double = 0) {
        
        guard let container = self.container else {
            fatalError()
        }
        
        guard let heightConstraint = self.container?.constraints.filter({$0.identifier == "height"}).first else {
            return
        }
        
        self.transform = CGAffineTransformMakeScale(0, 0)
        self.container?.layoutIfNeeded()
        
        self.alpha = 0
        
        heightConstraint.constant = CGFloat(distance)
        UIView.animateWithDuration(
            duration,
            delay: delay,
            usingSpringWithDamping: 0.7,
            initialSpringVelocity: 0,
            options: UIViewAnimationOptions.CurveLinear,
            animations: { () -> Void in
                container.layoutIfNeeded()
                self.transform = CGAffineTransformMakeScale(1.0, 1.0)
                self.alpha = 1
            }, completion: { (success) -> Void in
        })
    }
    
    public func hideAnimation(distance distance: Float, duration: Double, delay: Double = 0) {
        
        guard let container = self.container else {
            fatalError()
        }
        
        let heightConstraint = self.container?.constraints.filter {$0.identifier == "height"}.first
        
        guard heightConstraint != nil else {
            return
        }
        heightConstraint?.constant = CGFloat(distance)
        UIView.animateWithDuration(
            duration,
            delay: delay,
            options: UIViewAnimationOptions.CurveEaseIn,
            animations: { () -> Void in
                container.layoutIfNeeded()
                self.transform = CGAffineTransformMakeScale(0.01, 0.01)
            }, completion: { (success) -> Void in
                self.alpha = 0
                
                if let _ = self.container {
                    container.removeFromSuperview() // remove container
                }
        })
    }
    
    public func changeDistance(distance: CGFloat, animated: Bool, duration: Double = 0, delay: Double = 0) {
        
        guard let container = self.container else {
            fatalError()
        }
        
        let heightConstraint = self.container?.constraints.filter {$0.identifier == "height"}.first
        
        guard heightConstraint != nil else {
            return
        }
        
        heightConstraint?.constant = distance
        
        UIView.animateWithDuration(
            duration,
            delay: delay,
            options: UIViewAnimationOptions.CurveEaseIn,
            animations: { () -> Void in
                container.layoutIfNeeded()
            },
            completion: nil)
    }
    
    // MARK: layer animation
    
    public func rotationLayerAnimation(angle: Float, duration: Double) {
        if let aContainer = container {
            rotationLayerAnimation(aContainer, angle: angle, duration: duration)
        }
    }
}

extension UIView {
    
    public func rotationLayerAnimation(view: UIView, angle: Float, duration: Double) {
        
        let rotation = Init(CABasicAnimation(keyPath: "transform.rotation")) {
            $0.duration       = NSTimeInterval(duration)
            $0.toValue        = (angle.degrees)
            $0.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
        }
        view.layer.addAnimation(rotation, forKey: "rotation")
    }
}


CircleMenu/ViewController.swift
//
//  ViewController.swift
//  CircleMenu
//
//  Created by Alex K. on 27/01/16.
//  Copyright © 2016 Alex K. All rights reserved.
//

import UIKit

extension UIColor {
    static func color(red: Int, green: Int, blue: Int, alpha: Float) -> UIColor {
        return UIColor(
            colorLiteralRed: Float(1.0) / Float(255.0) * Float(red),
            green: Float(1.0) / Float(255.0) * Float(green),
            blue: Float(1.0) / Float(255.0) * Float(blue),
            alpha: alpha)
    }
}

class ViewController: UIViewController, CircleMenuDelegate {
    
    //    let colors = [UIColor.redColor(), UIColor.grayColor(), UIColor.greenColor(), UIColor.purpleColor()]
    let items: [(icon: String, color: UIColor)] = [
        ("icon_home", UIColor(red:0.19, green:0.57, blue:1, alpha:1)),
        ("icon_search", UIColor(red:0.22, green:0.74, blue:0, alpha:1)),
        ("notifications-btn", UIColor(red:0.96, green:0.23, blue:0.21, alpha:1)),
        ("settings-btn", UIColor(red:0.51, green:0.15, blue:1, alpha:1)),
        ("nearby-btn", UIColor(red:1, green:0.39, blue:0, alpha:1)),
    ]
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // add button
        
        //        let button = CircleMenu(
        //            frame: CGRect(x: 200, y: 200, width: 50, height: 50),
        //            normalIcon:"icon_menu",
        //            selectedIcon:"icon_close",
        //            buttonsCount: 4,
        //            duration: 4,
        //            distance: 120)
        //        button.backgroundColor = UIColor.lightGrayColor()
        //        button.delegate = self
        //        button.layer.cornerRadius = button.frame.size.width / 2.0
        //        view.addSubview(button)
        
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
    // MARK: <CircleMenuDelegate>
    
    func circleMenu(circleMenu: CircleMenu, willDisplay button: CircleMenuButton, atIndex: Int) {
        button.backgroundColor = items[atIndex].color
        button.setImage(UIImage(imageLiteral: items[atIndex].icon), forState: .Normal)
        
        // set highlited image
        let highlightedImage  = UIImage(imageLiteral: items[atIndex].icon).imageWithRenderingMode(.AlwaysTemplate)
        button.setImage(highlightedImage, forState: .Highlighted)
        button.tintColor = UIColor.init(colorLiteralRed: 0, green: 0, blue: 0, alpha: 0.3)
    }
    
    func circleMenu(circleMenu: CircleMenu, buttonWillSelected button: CircleMenuButton, atIndex: Int) {
        print("button will selected: \(atIndex)")
    }
    
    func circleMenu(circleMenu: CircleMenu, buttonDidSelected button: CircleMenuButton, atIndex: Int) {
        print("button did selected: \(atIndex)")
    }
}
